src/mudlet-lua/lua/Other.lua

```diff
@@ -700,3 +700,111 @@ function condenseMapLoad()
 
   return loadtime
 end
+
+do
+  -- management things
+
+  -- Dictionary with events as keys and lists of lua functions as values to dispatch events
+  -- to the right functions.
+  local handlers = {}
+
+  -- Remember highest hander ID to avoid ID reuse.
+  local highestHandlerId = 0
+  -- Helps us finding the right event handler from an ID.
+  local handlerIdsToHandlers = {}
+
+  -- C functions that get overwritten.
+  local origRegisterAnonymousEventHandler = registerAnonymousEventHandler
+
+  function registerAnonymousEventHandler(event, func, isOneShot)
+    if type(event) ~= "string"then
+      error(
+        string.format(
+          "Unexpected argument type in function registerAnonymousEventHandler for argument #1. String expected, got %s.",
+          type(event)
+        )
+      )
+    end
+
+    if type(func) ~= "function" and type(func) ~= "string" then
+      error(
+        string.format(
+          "Unexpected argument type in function registerAnonymousEventHandler for argument #2. Function or string expected, got %s.",
+          type(func)
+        )
+      )
+    end
+
+    if type(func) == "string" then
+      func = assert(loadstring(string.format("%s(...)", func)))
+    end
+
+    local eventHandlerId
+    if isOneShot then
+      -- wrap the original function to remove itself from the event handler list.
+      local origFunc = func
+      func = function(...)
+        origFunc(...)
+        killAnonymousEventHandler(eventHandlerId)
+      end
+    end
+
+    local existinghandlers = handlers[event]
+    if not existinghandlers then
+      existinghandlers = {}
+      handlers[event] = existinghandlers
+      origRegisterAnonymousEventHandler(event, "dispatchEventToFunctions")
+    end
+    local newId = #existinghandlers + 1
+    existinghandlers[newId] = func
+    -- Above may fill gaps if handlers have been deleted, but that's okay.
+    highestHandlerId = highestHandlerId + 1
```

Comment:
Isn't this flawed - if `local newId = #existinghandlers + 1` does indeed fill in a gap in the lua indexed array then `highestHandlerId` does not need to be incremented - or have I got the wrong end of (or even the wrong) stick? 😕

Reply:
`highestHandlerId` is not the highest index in the handler list for a specific event. Instead, it assigns global event handler IDs to a table that contains information about how to find this handler again. And since we don't want to re-use handlers, we have to increase that number, even though we may fill gaps.

This is what happens:
1. New handler gets created with the ID 1 and the find information `{event = "TestEvent", index = 1}`, meaning we can find the event handler with the ID 1 at `handlers["TestEvent"][1]`
2. New handler gets created with the ID 2 and the find information `{event = "gmcp.Char.Status", index = 1}`, meaning we can find the event handler with the ID 2 at `handlers["gmcp.Char.Status"][1]`
3. New handler gets created with the ID 3 and the find information `{event = "gmcp.Char.Status", index = 2}`, meaning we can find the event handler with the ID 3 at `handlers["gmcp.Char.Status"][2]`
4. Handler with ID 2 gets removed. Now the find information table and the handlers table look like this:
```lua
{
  {event = "TestEvent", index = 1},
  nil,
  {event = "gmcp.Char.Status", index = 2}
}

{
  TestEvent = {
    function() print("First event handler") end
  },
  ["gmcp.Char.Status"] = {
    nil,
    function() print("Third event handler") end
  }
}
```
5. New handler gets created with the ID 4 (remember we don't re-use IDs). The find information is now `{event = "gmcp.Char.Status", index = 1}`, meaning we can find the event handler with the ID 4 at `handlers["gmcp.Char.Status"][1]`. The find information table and the handler table look like this:
```lua
{
  {event = "TestEvent", index = 1},
  nil,
  {event = "gmcp.Char.Status", index = 2},
  {event = "gmcp.Char.Status", index = 1}
}

{
  TestEvent = {
    function() print("First event handler") end
  },
  ["gmcp.Char.Status"] = {
    function() print("Fourth event handler") end,
    function() print("Third event handler") end
  }
}
```

I hope this clears the confusion.

Reply:
Sort of! Thanks for the information - and the direction to a section in the _FM_ above... 🙂